packages/langium/src/workspace/profiler.ts

```diff
@@ -0,0 +1,189 @@
+            this.activeCategories.clear();
+        } else {
+            categories.forEach(category => this.activeCategories.delete(category));
+        }
+    }
+    createTask(category: ProfilingCategory, taskId: string): ProfilingTask {
+        if (!this.isActive(category)) {
+            throw new Error(`Category "${category}" is not active.`);
+        }
+        console.log(`Creating profiling task for '${category}.${taskId}'.`);
+        return new ProfilingTask((record: ProfilingRecord) => this.records.add(category, this.dumpRecord(category, record)), taskId);
+    }
+
+    private dumpRecord(category: string, record: ProfilingRecord): ProfilingRecord {
+        console.info(`Task ${category}.${record.identifier} executed in ${record.duration.toFixed(2)}ms and ended at ${record.date.toISOString()}`);
+
+        const result: Array<{ name: string, count: number, duration: number }> = [];
+        for (const key of record.entries.keys()) {
+            const values = record.entries.get(key);
+            const duration = values.reduce((p, c) => p + c);
+            result.push({ name: `${record.identifier}.${key}`, count: values.length, duration: duration });
+        }
+
+        // sum all duration
+        const taskInternalDuration = record.duration - result.map(r => r.duration).reduce((a, b) => a + b, 0);
+
+        result.push({ name: record.identifier, count: 1, duration: taskInternalDuration });
+
+        result.sort((a, b) => b.duration - a.duration);
+        function Round(value: number) { return Math.round(100 * value) / 100; }
+        console.table(result.map(e => { return { Element: e.name, Count: e.count, 'Self %': Round(100 * e.duration / record.duration), 'Time (ms)': Round(e.duration) }; }));
+        return record;
+    }
+    getRecords(...categories: ProfilingCategory[]): Stream<ProfilingRecord> {
+        if (categories.length === 0) {
+            // return all records
+            return this.records.values();
+        } else {
+            // return records for the given categories
+            return this.records.entries().filter((e) => categories.some(c => c === e[0])).flatMap(e => e[1]);
+        }
+    }
+    protected readonly records: MultiMap<string, ProfilingRecord> = new MultiMap();
+}
+
+interface ProfilingRecord {
+    // the record identifier (e.g: the grammar name)
+    identifier: string
+    // the date at which the record is generated
+    date: Date
+    // the duration of the record
+    duration: number
+    // a list of sub-tasks(string) called during the recording with
+    // for each sub-task the duration of each call.
+    entries: MultiMap<string, number>
+}
+
+export class ProfilingTask {
+    private startTime?: number;
+    private readonly addRecord: (record: ProfilingRecord) => void;
+
+    private readonly identifier: string;
+    private readonly stack: Array<{ id: string, start: number, content: number }> = [];
+    private readonly entries = new MultiMap<string, number>();
```

Comment:
Minor: I know that this is essentially an implementation detail of the `DefaultLangiumProfiler`, but just for consistency, it would be great if we could also keep everything `protected` in here.